src/device/asgard_kronmult_cpu.cpp

```diff
@@ -5,347 +5,770 @@
+      y[iy] = 0;
+    else if constexpr (beta_case == scalar_case::neg_one)
+      y[iy] = -y[iy];
+    else if constexpr (beta_case == scalar_case::other)
+      y[iy] *= beta;
+
+    // ma is the starting index of the operators for this y
+    int ma = iy * num_cols * num_terms * dimensions;
+
+    for (int jx = 0; jx < num_cols; jx++)
     {
-      int const i = iy * output_stride + stride;
-      T totalA    = 1;
-      for (int j = 0; j < dimensions; j++)
+      for (int t = 0; t < num_terms; t++)
       {
-        totalA *= pA[dimensions * i + j][0];
+        T totalA = 1;
+        for (int d = 0; d < dimensions; d++)
+          totalA *= vA[iA[ma++]];
+
+        if constexpr (alpha_case == scalar_case::one)
+          y[iy] += totalA * x[jx];
+        else if constexpr (alpha_case == scalar_case::neg_one)
+          y[iy] -= totalA * x[jx];
+        else
+          y[iy] += alpha * totalA * x[jx];
       }
-      pY[i][0] += totalA * pX[i][0];
     }
   }
 }
 
-template void run_cpu_variant0<float>(int const, float const *const[],
-                                      float const *const[], float *[],
-                                      int const, int const);
-template void run_cpu_variant0<double>(int const, double const *const[],
-                                       double const *const[], double *[],
-                                       int const, int const);
-
-template<typename T, int dimensions, int n>
-void run_cpu_variant(T const *const pA[], int const lda, T const *const pX[],
-                     T *pY[], int const num_batch, int const output_stride)
+/*!
+ * \brief Baseline kronmult algorithm on the CPU (dense case).
+ *
+ * The template handles up to 6D problems and could handle arbitrary n,
+ * however, only n = 2, 3, 4 is instantiated in asgard_kronmult_cpu.cpp.
+ * Fixing n as a template parameter allows for automatic unroll of
+ * the for-loop of the general algorithm, but also significantly
+ * increases the compile times.
+ *
+ * \tparam T is either float or double
+ * \tparam dimensions is between 1 and 6
+ * \tparam n is the size of the small matrices
+ * \tparam alpha_case must be one if alpha is 1 and neg_one if alpha is -1,
+ *         otherwise it must be scalar_case::other
+ *         alpha_case cannot be scalar_case::zero since that means
+ *         no multiplication
+ * \tparam beta_case must match beta, one for beta = 1, neg_one for beta = -1,
+ *         zero for beta = 0 and other in all other cases
+ */
+template<typename T, int dimensions, int n, scalar_case alpha_case,
+         scalar_case beta_case>
+void cpu_dense(int const num_rows, int num_cols, int const num_terms,
+               int const iA[], T const vA[], T const alpha, T const x[],
+               T const beta, T y[])
 {
   static_assert(1 <= dimensions and dimensions <= 6);
   static_assert(n > 1, "n must be positive and n==1 is a special case handled "
                        "by another method");
 
```

Comment:
Here is one more spot to suppress the warning.
```suggestion
  (void)alpha;
  (void)beta;
```